api/validation.go

```diff
@@ -633,19 +633,48 @@ func (e *EverestServer) validateDatabaseClusterOnCreate(
 			return errors.Join(errInsufficientPermissions, errors.New("missing permission to take backups"))
 		}
 	}
-	// To be able to create a cluster from a backup (restore backup to new cluster), the user needs to explicitly
-	// have permissions to take restores.
-	sourceBackup := pointer.Get(pointer.Get(pointer.Get(databaseCluster.Spec).DataSource).DbClusterBackupName)
-	if sourceBackup != "" {
-		if can, err := e.canRestore(user, namespace+"/"+sourceBackup); err != nil {
-			return err
-		} else if !can {
-			return errors.Join(errInsufficientPermissions, errors.New("missing permission to restore"))
-		}
+
+	if ok, err := e.enforceRestoreToNewDBRBAC(ctx.Request().Context(), user, namespace, databaseCluster); err != nil {
+		return err
+	} else if !ok {
+		return errors.Join(errInsufficientPermissions, errors.New("missing permission to take backups"))
 	}
 	return nil
 }
 
+// To be able to restore a backup to a new cluster, the following permissions are needed:
+// - create restores.
+// - read database cluster credentials.
+func (e *EverestServer) enforceRestoreToNewDBRBAC(
+	ctx context.Context, user, namespace string, databaseCluster *DatabaseCluster,
+) (bool, error) {
+	sourceBackup := pointer.Get(pointer.Get(pointer.Get(databaseCluster.Spec).DataSource).DbClusterBackupName)
+	if sourceBackup == "" {
+		return true, nil
+	}
+
+	if can, err := e.canRestore(user, namespace+"/"); err != nil {
+		return false, err
+	} else if !can {
+		return false, nil
+	}
+
+	// Get the name of the source database cluster.
+	bkp, err := e.kubeClient.GetDatabaseClusterBackup(ctx, namespace, sourceBackup)
+	if err != nil {
+		return false, errors.Join(err, errors.New("failed to get database cluster backup"))
+	}
```

Comment:
I know it's more or less implied (it would be hard to know the backup name without it) but shouldn't we also validate if the user has read access to that backup?

Reply:
I don't know if this is not overdoing stuff... One could argue that if you check for read permissions on the backup you should also have read permissions on the backup storage where that backup is stored. Am I jumping onto a rabbit hole?

Reply:
Updated